### PR TITLE
Mmmae dataset integration

### LINKED ISSUE #95: D4.2.2.6 Demonstrate Loading data from NetCDF datasets applied to spatio-temporal data
Current assumption that NerCDF dataset class will be useable for both static and spatio-temporal datasets. Demonstrate that it correclty loads and correctly dimensionalises the xarrays from both.

### LINKED ISSUE #98: D4.2.2.9. Pre-training demonstrator for ViT autoencoder
For a start, can run on one static layer and a subset of one ST dataset for the sake of demo.
For task completion, should be applied to multiple spatial and two spatio-temporal datasets.
Must implement MAE-reconstruction cost in training and validation steps.

### LINKED ISSUE #105: D4.2.2.11. Work out how to combine multiple datasets
Soil properties autoencoder combines data from multiple datasets in different modalilites. We need to decide where they can be combined to optimise memory use, handling of different modalities separately (for example, to create embeddings), and compatibility with pytorch-lightning and hydra.

There are two alternative approaches.

1. Create a single datamodule that would combine all static layers and a single datamodule for ST data. These would be used in different autoencoders. A single datamodule will handle:
-  getting the data from different local files in different models,
- verifying that they are in the same crs, 
- potentially regridding data to a single resolution, 
- tiling and all the transforms
- dataset splitting into train/test/val
- creation of a single dataloader
A single config file would be required for such module that would point to multiple dataset directories. The lightning trainer will take a single datamodule as is curently done in demos.

2. Create a datamodule per dataset format. Then, mulptiple hydra data config files are needed, each pointing to an individual dataset. The lightning trainer will have to handle multiple datamodules to train a single model.

### PR DESCRIPTION
**Pull request by MattArran**
## Description

Introduces and implements a `CorrespondingTilesDataset` class that can integrate multiple tiled spatial, temporal, or spatiotemporal datasets, providing data from the corresponding tiles in each dataset. The correspondence allows for different coordinate names and coordinate reference systems, and the data are in a form suitable for the Multimodal Masked Autoencoder, which has been updated for application to datasets with different dimensions.

**Breaking change**: In config files that instantiate `ToValuesBoundsTransform` and `ToValuesLocationsTransform`, `coords` needs updating to `dims`.

## Related Issue

Permits training of a vision transformer-based autoencoder on combinations of spatial, temporal, and spatiotemporal datasets, including those saved in NetCDF format. Thereby closes #95, closes #98, closes #105.

## Type of Change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] Test update (adding or modifying tests)

## Changes Made

- Refactors for better code style:
  - Makes `test_transforms.py` imports explicit.
  - Fixes whitespace in `transforms.py` and `test_transforms.py`.
  - Moves `ToValuesBoundsTransform` and `ToValuesLocationsTransform` classes into alphabetical order in `transforms.py`.
  - **Breaking change**: `coords` input of `ToValuesBoundsTransform` and `ToValuesLocationsTransform` class instantiation changed to `dims`, for consistency with `xarray.DataArray` and the rest of the code.
  - Replaces `assert` with `ValueError` input verification in `embedders.py` and `mmmae.py`.
- Adds support for different coordinate systems:
  - Introduces `CRS_spec`, `CRS_conversion_spec` and `CRS_convertor` classes in `transform_utils.py` and associated unit test in `test_transforms.py`, for coordinate reference system specification and conversion of xarray coordinates.
  - Adds `AddFixedCoordinates` transform class to `transforms.py`, to permit fixed-position information specification, with test in `test_transforms.py`.
  - Adds support for input dimension specification, via dictionaries, in instantiation of `MultimodalMaskedAutoencoder` and `PatchEmbed`-derived classes in `mmmae.py` and `embedders.py`, allowing inputs to be located in spaces other than those spanned by their axes. *Note*: the MMMAE requires inputs to be in the same  order as in the `input_shapes` dict and location or bounds coordinates to be in the same order as in the `position_space` dict. 
- Implements dataset combination
  - Adds `ZipDataset` class in new file `src/FRAME_FM/datasets/combined_dataset.py`, which acts equivalently to `zip` for `BaseDataset`s and allows multiple datasets to be provided to a dataloader.
  - Adds `CorrespondingTilesDataset` class in `combined_dataset.py`, which combines datasets by, for each tile T in the first dataset, using `TileIndexMapper` to identify the tile in each other dataset that contains T's centroid (or the closest edge tile, if T's centroid is outside the dataset).
  - Adds tests for each class in `tests/datasets/test_combined_dataset.py`.

## How Has This Been Tested?

- [x] Existing tests pass (`pytest`)
- [x] New tests added for changes
- [x] Tested manually (describe below)

Script `scripts/train_demonstrator` successfully runs at least one epoch on JASMIN.

## Checklist

- [x] My code follows the project's coding standards
- [x] I have performed a self-review of my code
- [x] I have added/updated tests that prove my fix or feature works
- [x] New and existing tests pass locally with my changes
- [x] I have updated documentation where necessary
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published

### CODE REVIEW

**Review by agstephens — APPROVED** [review]
Thanks @MattArran

**Comment by agstephens** [comment]
Hi @MattArran, please delete this branch if you are happy to do so.
